src/env_universal_common.rs

```diff
@@ -154,90 +131,76 @@ impl EnvUniversal {
-        // opened. If so, start over.
-        // 4. Read from the file. This can be elided if its dev/inode is unchanged since the last read
-        // 5. Open an adjacent temporary file
-        // 6. Write our changes to an adjacent file
-        // 7. Move the adjacent file into place via rename. This is assumed to be atomic.
-        // 8. Release the lock and close the file
-        //
-        // Consider what happens if Process 1 and 2 both do this simultaneously. Can there be data loss?
-        // Process 1 opens the file and then attempts to take the lock. Now, either process 1 will see
-        // the original file, or process 2's new file. If it sees the new file, we're OK: it's going to
-        // read from the new file, and so there's no data loss. If it sees the old file, then process 2
-        // must have locked it (if process 1 locks it, switch their roles). The lock will block until
-        // process 2 reaches step 7; at that point process 1 will reach step 2, notice that the file has
-        // changed, and then start over.
-        //
-        // It's possible that the underlying filesystem does not support locks (lockless NFS). In this
-        // case, we risk data loss if two shells try to write their universal variables simultaneously.
-        // In practice this is unlikely, since uvars are usually written interactively.
-        //
-        // Prior versions of fish used a hard link scheme to support file locking on lockless NFS. The
-        // risk here is that if the process crashes or is killed while holding the lock, future
-        // instances of fish will not be able to obtain it. This seems to be a greater risk than that of
-        // data loss on lockless NFS. Users who put their home directory on lockless NFS are playing
-        // with fire anyways.
         // If we have no changes, just load.
         if self.modified.is_empty() {
-            self.load_from_path_narrow(callbacks);
+            let callbacks = self.load_from_path_narrow();
             FLOG!(uvar_file, "universal log no modifications");
-            return false;
+            return (false, callbacks);
         }
 
-        let directory = wdirname(&self.vars_path).to_owned();
-
         FLOG!(uvar_file, "universal log performing full sync");
 
-        // Open the file.
-        let Some(mut vars_file) = self.open_and_acquire_lock() else {
-            FLOG!(uvar_file, "universal log open_and_acquire_lock() failed");
-            return false;
+        let rewrite = |old_file: &mut File,
+                       tmp_file: &mut File|
+         -> std::io::Result<Option<(u64, VarTable, CallbackDataList)>> {
+            let updates_from_old_file = self.load_from_file(old_file)?;
+            let vars = updates_from_old_file.as_ref().map(|(_, vars, _)| vars);
+            let var_ref = match vars {
+                Some(vars) => vars,
+                None => &self.vars,
+            };
```

Comment:
I think we can simplify this with `unwrap_or()`, and use shadowing (reducing the visibility of the owned `vars`).

```rust
let vars = vars..as_ref().unwrap_or(&self.vars);
```